### ISSUE TITLE
Bug: ClassNotFoundException, but it does exist in the project

### ISSUE THREAD

**Issue by BugsBunnyBR** [google/ksp#1038]
Hello there, I am trying to read my annotation using
```
classDeclaration.getAnnotationsByType(ActivityReference::class).first()
```

and getting this error

```
Caused by: com.google.devtools.ksp.KSTypeNotPresentException: java.lang.ClassNotFoundException: com.veepee.vpcore.route.BarImpl

```
This is a bit odd since the full qualified name in the exception seems to correct.

I tried [this](https://github.com/google/ksp/issues/888), but I still got the same error.

A project to reproduce it is attached [here](https://github.com/google/ksp/files/9077420/ksp-problem.zip)

To my specific needs I don't need to instantiate the annotation, but to get the String with the qualified name for `ActivityReference.activityName`, would it be possible with KSP ?

**Comment by BugsBunnyBR** [comment]
Ahhh, it seems that one MR to fix this already exists, https://github.com/google/ksp/pull/951

Sorry for the issue report.

**Comment by BugsBunnyBR** [comment]
Even after the merge of https://github.com/google/ksp/pull/1039 the code in my sample does not work well.

**Comment by BugsBunnyBR** [comment]
here we have an updated project using a local build

```
./gradlew -PkspVersion=1.7.20-dev-2726-1.0.7-snapshot publishToMavenLocal
```
and the build.gradle pointing to it
```
kotlin.code.style=official

kotlinVersion=1.7.20-dev-2726
kspVersion=1.7.20-dev-2726-1.0.7-snapshot
```

[ksp-problem.zip](https://github.com/google/ksp/files/9138752/ksp-problem.zip)

**Comment by CasualGitEnjoyer** [comment]
Did you resolve it?

**Comment by BugsBunnyBR** [comment]
No. I am still waiting the ksp developers to take a look at the sample project. Cc: @neetopia

**Comment by neetopia** [comment]
The problem with your project is that you are trying to get `BarImpl` which is declared in the source code of the workload `src`, therefore at KSP's runtime, `BarImpl` is not present on the classpath. Can you try to avoid using reflection if any of the KClass you want to access is not on classpath?

**Comment by unredundant** [comment]
Is this a planned feature for the future?  IMO there are totally valid use cases would be very convenient to reflect on elements in the classpath while using KSP.  

My particular use case is injecting permission policies defined by the developer into generated code.

**Comment by neetopia** [comment]
As I said, it is more like a limit of reflection itself (class has to be on classpath), is it possible to avoid using reflection for your use case? Or alternative way like having some interfaces in your library and reflect on the interface instead of the actual implementation generated?

**Comment by unredundant** [comment]
Not really, in this case I have an interface, which is implemented 
by the end user.  I would like the generated code to be able to introspect that implementation. 

But it sounds like I will need to rearchitect that, since it's a limitation of the language more than anything.

**Comment by neetopia** [comment]
By inspection do you mean generated code will need some logic that looks into the implementation provided by end users? It still sounds to me that reflection is not necessary, unless you need to access the object of the implementation at processing time. KSP already offers inspection into source code and generated code via KSP APIs, but if your use case is beyond the inspection provided by KSP, it might not work.

**Comment by unredundant** [comment]
personally, i don't need to access the object, I just need to be able yoink the fully qualified name, but the processor errors out when trying to access any information of the class

**Comment by BugsBunnyBR** [comment]
Me too, I just need to class name, I think this is super odd, because the ClassNotFoundException message that we receive contains the class name.

**Comment by JailedBird** [comment]
Me too, but I found that I could get the corresponding qualifiedName class from the Exception, howerve It's not a perfect solution;
```
import kotlin.reflect.KClass

@Target(AnnotationTarget.CLASS)
@Retention(AnnotationRetention.BINARY)
annotation class AutoService(val target: KClass<*> = Unit::class)
```
val spi: AutoService = element.findAnnotationWithType()
  

val targetInterfaceClassName = try {
    spi.target.qualifiedName.toString()
} catch (e: Exception) {
    /**
     * Bug: ksp: com.google.devtools.ksp.KSTypeNotPresentException: java.lang.ClassNotFoundException: cn.jailedbird.arouter.ksp.test.TestInterface
     * Official document: https://github.com/google/ksp/issues?q=ClassNotFoundException++KClass%3C*%3E
     * temporary fix method as follows, but it is not perfect!!!
     * TODO completely fix it!
     * */
    ((e as? KSTypeNotPresentException)?.cause as? ClassNotFoundException)?.message.toString()
}
```

**Comment by yuriisurzhykov** [comment]
Is there any updates for this issue? 
I'm having the same problem even for Any::class, when trying to get KClass from annotation property, it throws exception.
My KSP version is:  **1.9.21-1.0.15**

**Comment by JailedBird** [comment]
Recently, I adapted KSP for [WMRouter](https://github.com/meituan/WMRouter) project. During the adaptation process, I found a more suitable method to solve [issue1038](https://github.com/google/ksp/issues/1038)  (at least I think so)

We should adopt the solution proposed by bennyhuo [issue694](https://github.com/google/ksp/pull/694) to solve the problem of class parsing;



The operations of parsing the class object in the annotation and obtaining the fully qualified name of its class can be divided into the following types:

1、 Class[]：

Sample project：[JailedBird/WMRouterKspCompiler](https://github.com/JailedBird/WMRouterKspCompiler/blob/master/compiler/src/main/java/cn/jailedbird/wmrouter/ksp/compiler/RouterServiceSymbolProcessorProvider.kt)

```
@Target(ElementType.TYPE)
@Retention(RetentionPolicy.CLASS)
public @interface RouterService {
    Class[] interfaces();
}
```

solution：

```
val interfaceNames = WMRouterHelper.parseAnnotationClassParameter { service.interfaces.asList() }

fun parseAnnotationClassParameter(block: () -> List<KClass<*>>): List<String> {
    return try { // KSTypesNotPresentException will be thrown
        block.invoke().mapNotNull { it.qualifiedName }
    } catch (e: KSTypesNotPresentException) {
        val res = mutableListOf<String>()
        val ksTypes = e.ksTypes
        for (ksType in ksTypes) {
            val declaration = ksType.declaration
            if (declaration is KSClassDeclaration) {
                declaration.qualifiedName?.asString()?.let {
                    res.add(it)
                }
            }
        }
        res
    }
}
```



2、 Class （A little different from Class[]）

Sample project: [JailedBird/AutoService](https://github.com/JailedBird/AutoService/blob/main/compiler/src/main/java/cn/jailedbird/spi/compiler/AutoServiceSymbolProcessorProvider.kt)

```
@Target(AnnotationTarget.CLASS)
@Retention(AnnotationRetention.BINARY)
annotation class AutoService(val target: KClass<*> = Unit::class)
```

solution:

```
val targetInterfaceClassName = parseAnnotationClassParameter { spi.target }

@OptIn(KspExperimental::class)
fun parseAnnotationClassParameter(block: () -> KClass<*>): String? {
    return try { // KSTypeNotPresentException will be thrown
        block.invoke().qualifiedName
    } catch (e: KSTypeNotPresentException) {
        var res: String? = null
        val declaration = e.ksType.declaration
        if (declaration is KSClassDeclaration) {
            declaration.qualifiedName?.asString()?.let {
                res = it
            }
        }
        res
    }
}
```

**Comment by JailedBird** [comment]
> Is there any updates for this issue? I'm having the same problem even for Any::class, when trying to get KClass from annotation property, it throws exception. My KSP version is: **1.9.21-1.0.15**

Perhaps you can try this method：https://github.com/google/ksp/issues/1038#issuecomment-1958577350

**Comment by juliocbcotta** [comment]
This looks like a work around, not a fix or a reliable solution.

**Comment by yuriisurzhykov** [comment]
In addition, here is adapted solution to work with KotlinPoet:

```kotlin
@OptIn(KspExperimental::class)
fun parseAnnotationClassParameter(block: () -> List<KClass<*>>): List<ClassName> {
    return try {
        block.invoke().map { ClassName(it.java.`package`.name, it.simpleName.orEmpty()) }
    } catch (e: KSTypesNotPresentException) {
        val res = mutableListOf<ClassName>()
        val ksTypes = e.ksTypes
        for (ksType in ksTypes) {
            val declaration = ksType.declaration
            if (declaration is KSClassDeclaration) {
                res.add(declaration.toClassName())
            }
        }
        res
    }
}
```

**Comment by bubenheimer** [comment]
In terms of KSP API design, a critical takeaway from this issue should be to adjust the `getAnnotationsByType()` method (it's `@KspExperimental`, so within the realm of possibility). It currently does too much, leading users to automatically shoot themselves in the foot. The problem is that it returns an actual annotation class instance instead of `KSAnnotation`. The step of calling `toAnnotation()` on the result is essentially trivial and should be separate.

`getAnnotationsByType()` is clearly a useful API, but it is quite flawed to have it return `Sequence<T>` instead of `Sequence<KSAnnotation>`. This makes it unusable for general development, instead requiring users to duplicate just its first part to safely access arbitrary `KClass` annotation parameters. KSP provides no alternative API here, so everybody and their uncle gets hit by this issue.

https://github.com/google/ksp/blob/4ea5351aa154f1b3dd736781f80959a422749256/api/src/main/kotlin/com/google/devtools/ksp/utils.kt#L333-L339

@ting-yuan your thoughts?

At least provide alternative API capturing the essential parts.

**Comment by bubenheimer** [comment]
Nothing has changed in terms of relevance of this issue, AFAIK
